## Gathering

In [548]:
import pandas as pd
import numpy as np
import requests
import os
import tweepy
import json
from collections import Counter

#### - I have downloaded the csv files manually from udacitys resources tab on data wrangling session.

#### - I also apply  for developper account for Twitter for the purpose of getting Tweepy API data . 

#### - My @okyaysonmez account is now developper .I will focus on Twitter  API after DAND 's expected graduation

### 1.Twitter Archive Data

In [549]:
df_archieve = pd.read_csv("twitter-archive-enhanced.csv")

### 2.Image Prediction Data 

In [550]:
#I have created folder.
folder_name = 'Image_Predictions'
if not os.path.exists(folder_name):
    os.makedirs(folder_name)
    
#Getting the URL to get the image predictions tsv file.
url='https://d17h27t6h515a5.cloudfront.net/topher/2017/August/599fd2ad_image-predictions/image-predictions.tsv'
response=requests.get(url)

response.content

#To save the file to the computer
with open(os.path.join(folder_name, url.split('/')[-1]),mode='wb') as file:
        file.write(response.content)

In [551]:
image_predictions = pd.read_csv('Image_Predictions/image-predictions.tsv', sep = '\t')

### 3. Api Data (Retweet/Favourite count)

consumer_key = "iK9IgmuX16FfXaIrp70TQW0TR"
consumer_secret = "zg9IZ7RdMs0EwsgU9ZY7IPTkUCAm2Nz1hsGkxMIi3Bb6RODfv1"
access_token = "219020784-A73rjQmwhYJXTnJ8bo13bKvfjVTTDZfvebshSosh"
access_secret = "sQ1qOVeWTRr0vvj1HCfUjDSO6QChDm4DCuSx6Lm1CRkpo"

auth = tweepy.OAuthHandler(consumer_key, consumer_secret)
auth.set_access_token(access_token, access_secret)

api = tweepy.API(auth, wait_on_rate_limit = True, wait_on_rate_limit_notify = True)

tweet_ids = df_archieve.tweet_id.values

with open('tweet_json.txt', 'w') as output:
    for tweet_id in tweet_ids:
        try:
            tweet = api.get_status(tweet_id, tweet_mode='extended')
            json.dump(tweet._json, output)
            output.write('\n')
        except tweepy.TweepError:
            print('Error: tweet missing')

In [552]:
# Converting relevant JSON data to dataframe

tweets_list =[]

with open('tweet-json.txt') as json_file:
    for line in json_file:
    
        tweets_dict = {}
        tweets_json = json.loads(line)
        
        try:
            tweets_dict['tweet_id'] = tweets_json['extended_entities']['media'][0]['id']
        except:
            tweets_dict['tweet_id'] = 'na'

        tweets_dict['retweet_count'] = tweets_json['retweet_count']
        tweets_dict['favorite_count'] = tweets_json['favorite_count']
        
        tweets_list.append(tweets_dict)

In [553]:
tweets_summary = pd.DataFrame(tweets_list)

## Accessing

### df_archieve data

In [554]:
df_archieve.head()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   
1  892177421306343426                    NaN                  NaN   
2  891815181378084864                    NaN                  NaN   
3  891689557279858688                    NaN                  NaN   
4  891327558926688256                    NaN                  NaN   

                   timestamp  \
0  2017-08-01 16:23:56 +0000   
1  2017-08-01 00:17:27 +0000   
2  2017-07-31 00:18:03 +0000   
3  2017-07-30 15:58:51 +0000   
4  2017-07-29 16:00:24 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   
2  <a href="http://twitter.com/download/iphone" r...   
3  <a href="http://twitter.com/download/iphone" r...   
4  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   
1  This is Tilly. She's just checking pup on you....                  NaN   
2  This is Archie. He is a rare Norwegian Pouncin...                  NaN   
3  This is Darla. She commenced a snooze mid meal...                  NaN   
4  This is Franklin. He would like you to stop ca...                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   
1                       NaN                        NaN   
2                       NaN                        NaN   
3                       NaN                        NaN   
4                       NaN                        NaN   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   
1  https://twitter.com/dog_rates/status/892177421...                13   
2  https://twitter.com/dog_rates/status/891815181...                12   
3  https://twitter.com/dog_rates/status/891689557...                13   
4  https://twitter.com/dog_rates/status/891327558...                12   

   rating_denominator      name doggo floofer pupper puppo  
0                  10   Phineas  None    None   None  None  
1                  10     Tilly  None    None   None  None  
2                  10    Archie  None    None   None  None  
3                  10     Darla  None    None   None  None  
4                  10  Franklin  None    None   None  None

In [555]:
df_archieve.shape

(2356, 17)

In [556]:
df_archieve.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   tweet_id                    2356 non-null   int64  
 1   in_reply_to_status_id       78 non-null     float64
 2   in_reply_to_user_id         78 non-null     float64
 3   timestamp                   2356 non-null   object 
 4   source                      2356 non-null   object 
 5   text                        2356 non-null   object 
 6   retweeted_status_id         181 non-null    float64
 7   retweeted_status_user_id    181 non-null    float64
 8   retweeted_status_timestamp  181 non-null    object 
 9   expanded_urls               2297 non-null   object 
 10  rating_numerator            2356 non-null   int64  
 11  rating_denominator          2356 non-null   int64  
 12  name                        2356 non-null   object 
 13  doggo                       2356 

Tidy Data : Investigate the doggo,floofer,pupper,puppo columns, there should be only one column for these 4 information

In [557]:
df_archieve.doggo.unique(),df_archieve.floofer.unique(),df_archieve.pupper.unique(),df_archieve.puppo.unique()

(array(['None', 'doggo'], dtype=object),
 array(['None', 'floofer'], dtype=object),
 array(['None', 'pupper'], dtype=object),
 array(['None', 'puppo'], dtype=object))

In [558]:
df_archieve.doggo.value_counts(),df_archieve.floofer.value_counts(),df_archieve.pupper.value_counts(),df_archieve.puppo.value_counts()

(None     2259
 doggo      97
 Name: doggo, dtype: int64,
 None       2346
 floofer      10
 Name: floofer, dtype: int64,
 None      2099
 pupper     257
 Name: pupper, dtype: int64,
 None     2326
 puppo      30
 Name: puppo, dtype: int64)

Dirty_Data_1 :There is 745 records which name is none, and 55 records which name is "a". There are dummies as i suppose

In [559]:
df_archieve.name.value_counts()

None       745
a           55
Charlie     12
Oliver      11
Lucy        11
          ... 
Jonah        1
Jameson      1
Laela        1
Fynn         1
Kollin       1
Name: name, Length: 957, dtype: int64

Dirty_Data_2 : There are some records which rating_numerator/rating_denominator is more than 2

In [560]:
list_score = {}
for x in df_archieve.itertuples():
    try: 
        list_score[x.tweet_id] = x.rating_numerator/x.rating_denominator
    except:
        list_score[x.tweet_id] = 0    

In [561]:
sorted(list_score.values(),reverse = True)

[177.6,
 66.6,
 42.0,
 42.0,
 18.2,
 7.5,
 7.5,
 3.4285714285714284,
 2.7,
 2.6,
 1.7,
 1.5,
 1.5,
 1.4,
 1.4,
 1.4,
 1.4,
 1.4,
 1.4,
 1.4,
 1.4,
 1.4,
 1.4,
 1.4,
 1.4,
 1.4,
 1.4,
 1.4,
 1.4,
 1.4,
 1.4,
 1.4,
 1.4,
 1.4,
 1.4,
 1.4,
 1.4,
 1.4,
 1.4,
 1.4,
 1.4,
 1.4,
 1.4,
 1.4,
 1.4,
 1.4,
 1.4,
 1.4,
 1.4,
 1.4,
 1.4,
 1.4,
 1.4,
 1.4,
 1.4,
 1.4,
 1.4,
 1.4,
 1.4,
 1.4,
 1.4,
 1.4,
 1.4,
 1.4,
 1.4,
 1.4,
 1.4,
 1.3,
 1.3,
 1.3,
 1.3,
 1.3,
 1.3,
 1.3,
 1.3,
 1.3,
 1.3,
 1.3,
 1.3,
 1.3,
 1.3,
 1.3,
 1.3,
 1.3,
 1.3,
 1.3,
 1.3,
 1.3,
 1.3,
 1.3,
 1.3,
 1.3,
 1.3,
 1.3,
 1.3,
 1.3,
 1.3,
 1.3,
 1.3,
 1.3,
 1.3,
 1.3,
 1.3,
 1.3,
 1.3,
 1.3,
 1.3,
 1.3,
 1.3,
 1.3,
 1.3,
 1.3,
 1.3,
 1.3,
 1.3,
 1.3,
 1.3,
 1.3,
 1.3,
 1.3,
 1.3,
 1.3,
 1.3,
 1.3,
 1.3,
 1.3,
 1.3,
 1.3,
 1.3,
 1.3,
 1.3,
 1.3,
 1.3,
 1.3,
 1.3,
 1.3,
 1.3,
 1.3,
 1.3,
 1.3,
 1.3,
 1.3,
 1.3,
 1.3,
 1.3,
 1.3,
 1.3,
 1.3,
 1.3,
 1.3,
 1.3,
 1.3,
 1.3,
 1.3,
 1.3,
 1.3,
 1.3,
 1.3,
 1.3,
 1.3,
 1.3,
 1.3,
 1.3,
 

Dirty_Data3 : There are some records that have the denominator equals to 0 , this cause an error while calculating the ratio

In [562]:
df_archieve[df_archieve["rating_denominator"]==0]

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
313  835246439529840640           8.352460e+17           26259576.0   

                     timestamp  \
313  2017-02-24 21:54:03 +0000   

                                                source  \
313  <a href="http://twitter.com/download/iphone" r...   

                                                  text  retweeted_status_id  \
313  @jonnysun @Lin_Manuel ok jomny I know you're e...                  NaN   

     retweeted_status_user_id retweeted_status_timestamp expanded_urls  \
313                       NaN                        NaN           NaN   

     rating_numerator  rating_denominator  name doggo floofer pupper puppo  
313               960                   0  None  None    None   None  None

In [563]:
df_archieve.rating_denominator.min()

0

In [564]:
df_archieve.describe()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
count  2.356000e+03           7.800000e+01         7.800000e+01   
mean   7.427716e+17           7.455079e+17         2.014171e+16   
std    6.856705e+16           7.582492e+16         1.252797e+17   
min    6.660209e+17           6.658147e+17         1.185634e+07   
25%    6.783989e+17           6.757419e+17         3.086374e+08   
50%    7.196279e+17           7.038708e+17         4.196984e+09   
75%    7.993373e+17           8.257804e+17         4.196984e+09   
max    8.924206e+17           8.862664e+17         8.405479e+17   

       retweeted_status_id  retweeted_status_user_id  rating_numerator  \
count         1.810000e+02              1.810000e+02       2356.000000   
mean          7.720400e+17              1.241698e+16         13.126486   
std           6.236928e+16              9.599254e+16         45.876648   
min           6.661041e+17              7.832140e+05          0.000000   
25%           7.186315e+17              4.196984e+09         10.000000   
50%           7.804657e+17              4.196984e+09         11.000000   
75%           8.203146e+17              4.196984e+09         12.000000   
max           8.874740e+17              7.874618e+17       1776.000000   

       rating_denominator  
count         2356.000000  
mean            10.455433  
std              6.745237  
min              0.000000  
25%             10.000000  
50%             10.000000  
75%             10.000000  
max            170.000000

### image_predictions data

In [565]:
image_predictions.head()

tweet_id                                          jpg_url  \
0  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1  666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2  666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3  666044226329800704  https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4  666049248165822465  https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   

   img_num                      p1   p1_conf  p1_dog                  p2  \
0        1  Welsh_springer_spaniel  0.465074    True              collie   
1        1                 redbone  0.506826    True  miniature_pinscher   
2        1         German_shepherd  0.596461    True            malinois   
3        1     Rhodesian_ridgeback  0.408143    True             redbone   
4        1      miniature_pinscher  0.560311    True          Rottweiler   

    p2_conf  p2_dog                   p3   p3_conf  p3_dog  
0  0.156665    True    Shetland_sheepdog  0.061428    True  
1  0.074192    True  Rhodesian_ridgeback  0.072010    True  
2  0.138584    True           bloodhound  0.116197    True  
3  0.360687    True   miniature_pinscher  0.222752    True  
4  0.243682    True             Doberman  0.154629    True

In [566]:
image_predictions.tail()

tweet_id                                          jpg_url  \
2070  891327558926688256  https://pbs.twimg.com/media/DF6hr6BUMAAzZgT.jpg   
2071  891689557279858688  https://pbs.twimg.com/media/DF_q7IAWsAEuuN8.jpg   
2072  891815181378084864  https://pbs.twimg.com/media/DGBdLU1WsAANxJ9.jpg   
2073  892177421306343426  https://pbs.twimg.com/media/DGGmoV4XsAAUL6n.jpg   
2074  892420643555336193  https://pbs.twimg.com/media/DGKD1-bXoAAIAUK.jpg   

      img_num           p1   p1_conf  p1_dog                  p2   p2_conf  \
2070        2       basset  0.555712    True    English_springer  0.225770   
2071        1  paper_towel  0.170278   False  Labrador_retriever  0.168086   
2072        1    Chihuahua  0.716012    True            malamute  0.078253   
2073        1    Chihuahua  0.323581    True            Pekinese  0.090647   
2074        1       orange  0.097049   False               bagel  0.085851   

      p2_dog                           p3   p3_conf  p3_dog  
2070    True  German_short-haired_pointer  0.175219    True  
2071    True                      spatula  0.040836   False  
2072    True                       kelpie  0.031379    True  
2073    True                     papillon  0.068957    True  
2074   False                       banana  0.076110   False

In [567]:
image_predictions.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   tweet_id  2075 non-null   int64  
 1   jpg_url   2075 non-null   object 
 2   img_num   2075 non-null   int64  
 3   p1        2075 non-null   object 
 4   p1_conf   2075 non-null   float64
 5   p1_dog    2075 non-null   bool   
 6   p2        2075 non-null   object 
 7   p2_conf   2075 non-null   float64
 8   p2_dog    2075 non-null   bool   
 9   p3        2075 non-null   object 
 10  p3_conf   2075 non-null   float64
 11  p3_dog    2075 non-null   bool   
dtypes: bool(3), float64(3), int64(2), object(4)
memory usage: 152.1+ KB


### tweets_summary data

In [568]:
tweets_summary.head()

tweet_id  retweet_count  favorite_count
0  892420639486877696           8853           39467
1  892177413194625024           6514           33819
2  891815175371796480           4328           25461
3  891689552724799489           8964           42908
4  891327551943041024           9774           41048

In [569]:
tweets_summary.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2354 entries, 0 to 2353
Data columns (total 3 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   tweet_id        2354 non-null   object
 1   retweet_count   2354 non-null   int64 
 2   favorite_count  2354 non-null   int64 
dtypes: int64(2), object(1)
memory usage: 55.3+ KB


In [570]:
tweets_summary.describe()

retweet_count  favorite_count
count    2354.000000     2354.000000
mean     3164.797366     8080.968564
std      5284.770364    11814.771334
min         0.000000        0.000000
25%       624.500000     1415.000000
50%      1473.500000     3603.500000
75%      3652.000000    10122.250000
max     79515.000000   132810.000000

## Cleaning

First of all, i should copy the DataFrame in case i need to have the orijinal data in the future

In [571]:
df_archieve_clean = df_archieve.copy()
image_predictions_clean =image_predictions.copy()
tweets_summary_clean = tweets_summary.copy()


#### Tidy :

1. To define new columns named dog_type which have the doggo, floofer, pupper,puppo info into one column on df_achieve. 

2. Get only required columns from twitter-archive-enhanced.csv

3. Create breed & confidence on image prediction table

4. Get only required columns from image_predictions

#### Quality :
Table  : twitter-archive-enhanced :

1. Delete the 745 records which name are Blank &  55 records whih aredummy "a" on 
2. Update 11 records which have the ratio rating_numerator/ rating_denominator > 2
3. Delete the tweet that have the tweet_id = 835246439529840640 because it has denominator = 0 causes zero division error
4. Change the timestamp adatatypes to datetime, It is str (object)

Table : image-predictions 

5. Delete "-" records between words. update the words at he same format at image-predictions table

Table : tweet-summary

6. Delete  the records that have twitter_id = "na"

7. Tweet id type should be integer instead of string


.
8. Datasets in separate dataframes. Combine 3 dfs (tweet id will be used)


### Tidy Issues


#### DEFINE :  1. To define new columns named stage which have the doggo, floofer, pupper,puppo info into one column on df_achieve

CODE :

In [572]:
df_archieve_clean.head(1)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   

                   timestamp  \
0  2017-08-01 16:23:56 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   

   rating_denominator     name doggo floofer pupper puppo  
0                  10  Phineas  None    None   None  None

In [573]:
dog_type = []

string_in = ['puppo', 'pupper', 'doggo', 'floof']
string_out = ['puppo', 'pupper', 'doggo', 'floofer']

for row in df_archieve_clean['text']:
    row = row.lower()
    for word in string_in:
        if word in str(row):
            dog_type.append(string_out[string_in.index(word)])
            break
    else:
        dog_type.append('None')

df_archieve_clean['dog_type'] = dog_type
df_archieve_clean['dog_type'].value_counts()


None       1905
pupper      283
doggo        93
puppo        38
floofer      37
Name: dog_type, dtype: int64

In [574]:
df_archieve_clean.drop(['doggo', 'floofer', 'pupper', 'puppo'], axis=1, inplace=True)

TEST :

In [575]:
df_archieve_clean.head(1)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   

                   timestamp  \
0  2017-08-01 16:23:56 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   

   rating_denominator     name dog_type  
0                  10  Phineas     None

#### DEFINE : 2. Get only required columns from twitter-archive-enhanced.csv

#### CODE :

In [576]:
df_archieve_clean.drop(['in_reply_to_status_id', 'in_reply_to_user_id', 'retweeted_status_id', 'retweeted_status_user_id','retweeted_status_timestamp','expanded_urls'], axis=1, inplace=True)

In [577]:
df_archieve_clean.drop(['source'], axis=1, inplace=True)

#### TEST : 

In [578]:
df_archieve_clean

tweet_id                  timestamp  \
0     892420643555336193  2017-08-01 16:23:56 +0000   
1     892177421306343426  2017-08-01 00:17:27 +0000   
2     891815181378084864  2017-07-31 00:18:03 +0000   
3     891689557279858688  2017-07-30 15:58:51 +0000   
4     891327558926688256  2017-07-29 16:00:24 +0000   
...                  ...                        ...   
2351  666049248165822465  2015-11-16 00:24:50 +0000   
2352  666044226329800704  2015-11-16 00:04:52 +0000   
2353  666033412701032449  2015-11-15 23:21:54 +0000   
2354  666029285002620928  2015-11-15 23:05:30 +0000   
2355  666020888022790149  2015-11-15 22:32:08 +0000   

                                                   text  rating_numerator  \
0     This is Phineas. He's a mystical boy. Only eve...                13   
1     This is Tilly. She's just checking pup on you....                13   
2     This is Archie. He is a rare Norwegian Pouncin...                12   
3     This is Darla. She commenced a snooze mid meal...                13   
4     This is Franklin. He would like you to stop ca...                12   
...                                                 ...               ...   
2351  Here we have a 1949 1st generation vulpix. Enj...                 5   
2352  This is a purebred Piers Morgan. Loves to Netf...                 6   
2353  Here is a very happy pup. Big fan of well-main...                 9   
2354  This is a western brown Mitsubishi terrier. Up...                 7   
2355  Here we have a Japanese Irish Setter. Lost eye...                 8   

      rating_denominator      name dog_type  
0                     10   Phineas     None  
1                     10     Tilly     None  
2                     10    Archie     None  
3                     10     Darla     None  
4                     10  Franklin     None  
...                  ...       ...      ...  
2351                  10      None     None  
2352                  10         a     None  
2353                  10         a     None  
2354                  10         a     None  
2355                  10      None     None  

[2356 rows x 7 columns]

#### DEFINE : 3. Create breed & confidence on image prediction table

#### CODE

In [579]:
breed = []
confidence = []

def breed_confidence(row):
    if row['p1_dog'] == True:
        breed.append(row['p1'])
        confidence.append(row['p1_conf'])
    elif row['p2_dog'] == True:
        breed.append(row['p2'])
        confidence.append(row['p2_conf'])
    elif row['p3_dog'] == True:
        breed.append(row['p3'])
        confidence.append(row['p3_conf'])
    else:
        breed.append('Unidentifiable')
        confidence.append(0)
      

image_predictions_clean.apply(breed_confidence, axis=1)
image_predictions_clean['breed'] = breed
image_predictions_clean['confidence'] = confidence

#### TEST

In [580]:
image_predictions_clean.head()

tweet_id                                          jpg_url  \
0  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1  666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2  666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3  666044226329800704  https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4  666049248165822465  https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   

   img_num                      p1   p1_conf  p1_dog                  p2  \
0        1  Welsh_springer_spaniel  0.465074    True              collie   
1        1                 redbone  0.506826    True  miniature_pinscher   
2        1         German_shepherd  0.596461    True            malinois   
3        1     Rhodesian_ridgeback  0.408143    True             redbone   
4        1      miniature_pinscher  0.560311    True          Rottweiler   

    p2_conf  p2_dog                   p3   p3_conf  p3_dog  \
0  0.156665    True    Shetland_sheepdog  0.061428    True   
1  0.074192    True  Rhodesian_ridgeback  0.072010    True   
2  0.138584    True           bloodhound  0.116197    True   
3  0.360687    True   miniature_pinscher  0.222752    True   
4  0.243682    True             Doberman  0.154629    True   

                    breed  confidence  
0  Welsh_springer_spaniel    0.465074  
1                 redbone    0.506826  
2         German_shepherd    0.596461  
3     Rhodesian_ridgeback    0.408143  
4      miniature_pinscher    0.560311

#### DEFINE : 4. Get only required columns from image_predictions

CODE : 

In [581]:
image_predictions_clean.drop(['jpg_url','img_num','p1','p1_conf','p1_dog','p2','p2_conf','p2_dog','p3','p3_conf','p3_dog'], axis=1, inplace=True)

TEST :

In [582]:
image_predictions_clean.head()

tweet_id                   breed  confidence
0  666020888022790149  Welsh_springer_spaniel    0.465074
1  666029285002620928                 redbone    0.506826
2  666033412701032449         German_shepherd    0.596461
3  666044226329800704     Rhodesian_ridgeback    0.408143
4  666049248165822465      miniature_pinscher    0.560311

### QUALITY ISSUES

Table  : twitter-archive-enhanced :

#### Define : 1.Delete the 745 records which name are Blank & 55 records whih are dummy "a" and so on. 
There are many. We should delete all of them

Code:

In [583]:
no_names = ['a', 'an', 'the', 'just', 'one', 'very', 'quite', 'not', 'actually', 
             'mad', 'space', 'infuriating', 'all', 'officially', '0', 'old', 'life',
             'unacceptable', 'my', 'incredibly', 'by', 'his', 'such']

for name in df_archieve_clean.name:
    if name in no_names:
        df_archieve_clean.name[df_archieve_clean.name == name] = 'None'


/Users/okyay/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:7: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  import sys


Test: 

In [584]:
df_archieve_clean[df_archieve_clean["name"]=="an"]

Empty DataFrame
Columns: [tweet_id, timestamp, text, rating_numerator, rating_denominator, name, dog_type]
Index: []

In [585]:
df_archieve_clean.head()

tweet_id                  timestamp  \
0  892420643555336193  2017-08-01 16:23:56 +0000   
1  892177421306343426  2017-08-01 00:17:27 +0000   
2  891815181378084864  2017-07-31 00:18:03 +0000   
3  891689557279858688  2017-07-30 15:58:51 +0000   
4  891327558926688256  2017-07-29 16:00:24 +0000   

                                                text  rating_numerator  \
0  This is Phineas. He's a mystical boy. Only eve...                13   
1  This is Tilly. She's just checking pup on you....                13   
2  This is Archie. He is a rare Norwegian Pouncin...                12   
3  This is Darla. She commenced a snooze mid meal...                13   
4  This is Franklin. He would like you to stop ca...                12   

   rating_denominator      name dog_type  
0                  10   Phineas     None  
1                  10     Tilly     None  
2                  10    Archie     None  
3                  10     Darla     None  
4                  10  Franklin     None

#### Define : 2.Update 11 records which have the ratio rating_numerator/ rating_denominator > 2

Code:

In [586]:
 df_11 = df_archieve_clean[df_archieve_clean.rating_numerator/df_archieve_clean.rating_denominator >2]

In [587]:
df_11["rating_numerator"]=df_11["rating_denominator"]

/Users/okyay/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


In [588]:
df_archieve_clean = df_archieve_clean.drop(df_archieve_clean[df_archieve_clean.rating_numerator/df_archieve_clean.rating_denominator >2].index)


In [589]:
df_archieve_clean = df_archieve_clean.append(df_11)

Test:

In [590]:
df_archieve_clean[df_archieve_clean.rating_numerator/df_archieve_clean.rating_denominator >2]

Empty DataFrame
Columns: [tweet_id, timestamp, text, rating_numerator, rating_denominator, name, dog_type]
Index: []

In [591]:
df_archieve_clean.head()

tweet_id                  timestamp  \
0  892420643555336193  2017-08-01 16:23:56 +0000   
1  892177421306343426  2017-08-01 00:17:27 +0000   
2  891815181378084864  2017-07-31 00:18:03 +0000   
3  891689557279858688  2017-07-30 15:58:51 +0000   
4  891327558926688256  2017-07-29 16:00:24 +0000   

                                                text  rating_numerator  \
0  This is Phineas. He's a mystical boy. Only eve...                13   
1  This is Tilly. She's just checking pup on you....                13   
2  This is Archie. He is a rare Norwegian Pouncin...                12   
3  This is Darla. She commenced a snooze mid meal...                13   
4  This is Franklin. He would like you to stop ca...                12   

   rating_denominator      name dog_type  
0                  10   Phineas     None  
1                  10     Tilly     None  
2                  10    Archie     None  
3                  10     Darla     None  
4                  10  Franklin     None

#### Define : 3.Delete the tweet that have the tweet_id = 835246439529840640 because it has denominator = 0 causes zero division error

Code:

In [592]:
df_archieve_clean = df_archieve_clean.drop(df_archieve_clean[df_archieve_clean["rating_denominator"]==0].index)

Test:

In [593]:
df_archieve_clean[df_archieve_clean["rating_denominator"]==0]

Empty DataFrame
Columns: [tweet_id, timestamp, text, rating_numerator, rating_denominator, name, dog_type]
Index: []

#### Define : 4.Change the timestamp datatypes to datetime, It is str (object)

In [594]:
df_archieve_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2355 entries, 0 to 2074
Data columns (total 7 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   tweet_id            2355 non-null   int64 
 1   timestamp           2355 non-null   object
 2   text                2355 non-null   object
 3   rating_numerator    2355 non-null   int64 
 4   rating_denominator  2355 non-null   int64 
 5   name                2355 non-null   object
 6   dog_type            2355 non-null   object
dtypes: int64(3), object(4)
memory usage: 147.2+ KB


CODE

In [595]:
df_archieve_clean.timestamp = pd.to_datetime(df_archieve_clean.timestamp)

TEST

In [596]:
df_archieve_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2355 entries, 0 to 2074
Data columns (total 7 columns):
 #   Column              Non-Null Count  Dtype              
---  ------              --------------  -----              
 0   tweet_id            2355 non-null   int64              
 1   timestamp           2355 non-null   datetime64[ns, UTC]
 2   text                2355 non-null   object             
 3   rating_numerator    2355 non-null   int64              
 4   rating_denominator  2355 non-null   int64              
 5   name                2355 non-null   object             
 6   dog_type            2355 non-null   object             
dtypes: datetime64[ns, UTC](1), int64(3), object(3)
memory usage: 147.2+ KB


Table : image-predictions 

#### DEFINE : 5. Delete "-" records between words. update the words at he same format at image-predictions table 

CODE

image_predictions_clean.breed.value_counts()

In [597]:
image_predictions_clean.breed  = image_predictions_clean.breed.str.replace("_"," ").str.title()

TEST

In [598]:
image_predictions_clean.breed.value_counts()

Unidentifiable          324
Golden Retriever        173
Labrador Retriever      113
Pembroke                 96
Chihuahua                95
                       ... 
Irish Wolfhound           1
Standard Schnauzer        1
Scotch Terrier            1
Japanese Spaniel          1
Bouvier Des Flandres      1
Name: breed, Length: 114, dtype: int64

#### DEFINE : 6. Delete  the records that have twitter_id = "na" from twitter_summary

CODE

In [599]:
tweets_summary_clean= tweets_summary_clean.drop(tweets_summary_clean[tweets_summary_clean["tweet_id"]=="na"].index)

TEST

In [600]:
tweets_summary_clean[tweets_summary_clean["tweet_id"]=="na"]

Empty DataFrame
Columns: [tweet_id, retweet_count, favorite_count]
Index: []

#### DEFINE 7 Tweet id type should be integer instead of string at Tweet Summary

CODE

In [601]:
tweets_summary_clean["tweet_id"] =pd.to_numeric(tweets_summary_clean["tweet_id"])

TEST

In [602]:
tweets_summary_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2073 entries, 0 to 2353
Data columns (total 3 columns):
 #   Column          Non-Null Count  Dtype
---  ------          --------------  -----
 0   tweet_id        2073 non-null   int64
 1   retweet_count   2073 non-null   int64
 2   favorite_count  2073 non-null   int64
dtypes: int64(3)
memory usage: 64.8 KB


#### DEFINE 8: Combine all dataframes (df_archieve_clean, image_predictions_clean ,tweets_summary_clean )

CODE

In [630]:
master_df_pt1 = pd.merge(df_archieve_clean, image_predictions_clean, on='tweet_id', how='left')


master_df = pd.merge(master_df_pt1, tweets_summary_clean,on='tweet_id',how='left')



TEST

In [631]:
master_df.head(40)

tweet_id                 timestamp  \
0   892420643555336193 2017-08-01 16:23:56+00:00   
1   892177421306343426 2017-08-01 00:17:27+00:00   
2   891815181378084864 2017-07-31 00:18:03+00:00   
3   891689557279858688 2017-07-30 15:58:51+00:00   
4   891327558926688256 2017-07-29 16:00:24+00:00   
5   891087950875897856 2017-07-29 00:08:17+00:00   
6   890971913173991426 2017-07-28 16:27:12+00:00   
7   890729181411237888 2017-07-28 00:22:40+00:00   
8   890609185150312448 2017-07-27 16:25:51+00:00   
9   890240255349198849 2017-07-26 15:59:51+00:00   
10  890006608113172480 2017-07-26 00:31:25+00:00   
11  889880896479866881 2017-07-25 16:11:53+00:00   
12  889665388333682689 2017-07-25 01:55:32+00:00   
13  889638837579907072 2017-07-25 00:10:02+00:00   
14  889531135344209921 2017-07-24 17:02:04+00:00   
15  889278841981685760 2017-07-24 00:19:32+00:00   
16  888917238123831296 2017-07-23 00:22:39+00:00   
17  888804989199671297 2017-07-22 16:56:37+00:00   
18  888554962724278272 2017-07-22 00:23:06+00:00   
19  888202515573088257 2017-07-21 01:02:36+00:00   
20  888078434458587136 2017-07-20 16:49:33+00:00   
21  887705289381826560 2017-07-19 16:06:48+00:00   
22  887517139158093824 2017-07-19 03:39:09+00:00   
23  887473957103951883 2017-07-19 00:47:34+00:00   
24  887343217045368832 2017-07-18 16:08:03+00:00   
25  887101392804085760 2017-07-18 00:07:08+00:00   
26  886983233522544640 2017-07-17 16:17:36+00:00   
27  886736880519319552 2017-07-16 23:58:41+00:00   
28  886680336477933568 2017-07-16 20:14:00+00:00   
29  886366144734445568 2017-07-15 23:25:31+00:00   
30  886267009285017600 2017-07-15 16:51:35+00:00   
31  886258384151887873 2017-07-15 16:17:19+00:00   
32  886054160059072513 2017-07-15 02:45:48+00:00   
33  885984800019947520 2017-07-14 22:10:11+00:00   
34  885528943205470208 2017-07-13 15:58:47+00:00   
35  885518971528720385 2017-07-13 15:19:09+00:00   
36  885311592912609280 2017-07-13 01:35:06+00:00   
37  885167619883638784 2017-07-12 16:03:00+00:00   
38  884925521741709313 2017-07-12 00:01:00+00:00   
39  884876753390489601 2017-07-11 20:47:12+00:00   

                                                 text  rating_numerator  \
0   This is Phineas. He's a mystical boy. Only eve...                13   
1   This is Tilly. She's just checking pup on you....                13   
2   This is Archie. He is a rare Norwegian Pouncin...                12   
3   This is Darla. She commenced a snooze mid meal...                13   
4   This is Franklin. He would like you to stop ca...                12   
5   Here we have a majestic great white breaching ...                13   
6   Meet Jax. He enjoys ice cream so much he gets ...                13   
7   When you watch your owner call another dog a g...                13   
8   This is Zoey. She doesn't want to be one of th...                13   
9   This is Cassie. She is a college pup. Studying...                14   
10  This is Koda. He is a South Australian decksha...                13   
11  This is Bruno. He is a service shark. Only get...                13   
12  Here's a puppo that seems to be on the fence a...                13   
13  This is Ted. He does his best. Sometimes that'...                12   
14  This is Stuart. He's sporting his favorite fan...                13   
15  This is Oliver. You're witnessing one of his m...                13   
16  This is Jim. He found a fren. Taught him how t...                12   
17  This is Zeke. He has a new stick. Very proud o...                13   
18  This is Ralphus. He's powering up. Attempting ...                13   
19  RT @dog_rates: This is Canela. She attempted s...                13   
20  This is Gerald. He was just told he didn't get...                12   
21  This is Jeffrey. He has a monopoly on the pool...                13   
22  I've yet to rate a Venezuelan Hover Wiener. Th...                14   
23  This is Canela. She attempted some fancy porch...                13   
24  Yo

In [629]:
master_df["retweet_count"] == Blank

NameError: name 'Blank' is not defined

#### Store

In [605]:
# Saving data to csv

folder_name = 'Final_Product'
if not os.path.exists(folder_name):
    os.makedirs(folder_name)
    
    
master_df.to_csv('Final_Product/twitter_archive_master_end.csv')
df_archieve_clean.to_csv('Final_Product/clean_twitter_archive_data_end.csv')
image_predictions_clean.to_csv('Final_Product/clean_image_prediction_data_end.csv')
tweets_summary_clean.to_csv('Final_Product/tweets_summary_end.csv')

#### Analysis

#### Visualizing